### DISCUSSION TITLE
Optimistic updates failing seemingly randomly

### DISCUSSION THREAD

**Discussion by gregoralbrecht** [TanStack/query#2580] [Q&A]
Hey there! 👋  I'm trying to get optimistic updates working with 2 custom hooks, but I must be doing something horribly wrong because some updates just don't appear like I'd expect them to. 

For the sake of easy reproduction, I tried finding a public dummy API that supports `POST` requests, so here's a sandbox: https://codesandbox.io/s/musing-keldysh-s2bmn?file=/src/hooks/useAddPost.ts

More specifically, I have 2 main issues with this:

### 1. Optimistic updates randomly disappear after being added
I'd expect this to just work fine and it does in like 90% of the time. However, sometimes, the optimistically-added post briefly appears in the list, then disappears again until it's refetched some time later. You can see this by adding some random posts and seeing the described behavior.

### 2. The query is randomly undefined / doesn't exist
At least in this CodeSandbox, the query is sometimes `undefined`. Is this due to it being a dev server? Hot module replacement? I honestly don't know, but sometimes `queryClient.getQueryData()` returns undefined (as mentioned in the docs). Not sure why that is, though, and I don't know how to handle that case for optimistic updates?

I'd much appreciate any help here! 🙃 

PS: I already looked at the ["Optimistic Updates in TypeScript"](https://react-query.tanstack.com/examples/optimistic-updates-typescript) example and it seems broken.

PPS: I also wanted to ask this on Discord, but the [current link](https://discord.com/invite/mQd7egN) results in a "Unable to accept invite"

**Comment by TkDodo** [comment]
this seems to be purely an issue with that json placeholder api. We create an entry, and the next fetch sometimes doesn't return that entry 🤷 .

> PS: I already looked at the "Optimistic Updates in TypeScript" example and it seems broken.

what's broken there please?

> PPS: I also wanted to ask this on Discord, but the current link results in a "Unable to accept invite"

hmm the discord link works for me...

**Reply by gregoralbrecht** [reply]
> seems to be purely an issue with that json placeholder API

Well, that's reassuring then :) Thanks for taking a look! 🙏 

> what's broken there please?

Sorry for the vague description, seems like I need some sleep. Everything works fine as long as you let CodeSandbox do its thing first.

Do you have any idea why the query cache might get unmounted randomly? https://github.com/tannerlinsley/react-query/discussions/2065 seems to mention the same issue.

**Reply by TkDodo** [reply]
I answered in that other discussion: https://github.com/tannerlinsley/react-query/discussions/2065#discussioncomment-1193262
